tempesta_fw/http_frame.c

```diff
@@ -484,43 +489,25 @@ tfw_h2_send_rst_stream(TfwH2Ctx *ctx, unsigned int id, TfwH2Err err_code)
 	return tfw_h2_send_frame(ctx, &hdr, &data);
 }
 
-static inline int
-tfw_h2_conn_terminate(TfwH2Ctx *ctx, TfwH2Err err_code)
-{
-	return tfw_h2_send_goaway(ctx, err_code);
-}
-
-static inline int
-tfw_h2_stream_terminate(TfwH2Ctx *ctx, unsigned int id, TfwStream **stream,
-			TfwH2Err err_code)
+void
+tfw_h2_conn_terminate_close(TfwH2Ctx *ctx, TfwH2Err err_code, bool close)
 {
-	if (stream && *stream) {
-		--ctx->streams_num;
-		tfw_h2_stop_stream(&ctx->sched, stream);
-	}
+	TfwH2Conn *conn = container_of(ctx, TfwH2Conn, h2);
 
-	return tfw_h2_send_rst_stream(ctx, id, err_code);
+	if (tfw_h2_send_goaway(ctx, err_code) && close)
```

Comment:
Yes, this is expected behavior, since in case of successful execution of `tfw_h2_send_goaway()` - message will be send to our `work_queue` and connections will be closed automatically due to set `SS_F_CONN_CLOSE` flags. If `close` is false, then the procedure is called from context of client connection (during request receiving on frame or http layers) and connection will be closed due returned T_DROP code up to `ss_tcp_process_data()`/`ss_tcp_data_ready()`.